File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/CommodityRelativeExpirationDates.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CommodityRelativeExpirationDates : ISerialized
    {
        public CommodityRelativeExpirationDates(XmlNode xmlNode)
        {
            XmlNode expireRelativeToEventNode = xmlNode.SelectSingleNode("expireRelativeToEvent");
            
            if (expireRelativeToEventNode != null)
            {
                if (expireRelativeToEventNode.Attributes["href"] != null || expireRelativeToEventNode.Attributes["id"] != null) 
                {
                    if (expireRelativeToEventNode.Attributes["id"] != null) 
                    {
                        expireRelativeToEventIDRef_ = expireRelativeToEventNode.Attributes["id"].Value;
                        CommodityExpireRelativeToEvent ob = new CommodityExpireRelativeToEvent(expireRelativeToEventNode);
                        IDManager.SetID(expireRelativeToEventIDRef_, ob);
                    }
                    else if (expireRelativeToEventNode.Attributes["href"] != null)
                    {
                        expireRelativeToEventIDRef_ = expireRelativeToEventNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expireRelativeToEvent_ = new CommodityExpireRelativeToEvent(expireRelativeToEventNode);
                    }
                }
                else
                {
                    expireRelativeToEvent_ = new CommodityExpireRelativeToEvent(expireRelativeToEventNode);
                }
            }
            
        
            XmlNode expirationDateOffsetNode = xmlNode.SelectSingleNode("expirationDateOffset");
            
            if (expirationDateOffsetNode != null)
            {
                if (expirationDateOffsetNode.Attributes["href"] != null || expirationDateOffsetNode.Attributes["id"] != null) 
                {
                    if (expirationDateOffsetNode.Attributes["id"] != null) 
                    {
                        expirationDateOffsetIDRef_ = expirationDateOffsetNode.Attributes["id"].Value;
                        DateOffset ob = new DateOffset(expirationDateOffsetNode);
                        IDManager.SetID(expirationDateOffsetIDRef_, ob);
                    }
                    else if (expirationDateOffsetNode.Attributes["href"] != null)
                    {
                        expirationDateOffsetIDRef_ = expirationDateOffsetNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationDateOffset_ = new DateOffset(expirationDateOffsetNode);
                    }
                }
                else
                {
                    expirationDateOffset_ = new DateOffset(expirationDateOffsetNode);
                }
            }
            
        
            XmlNode businessCentersReferenceNode = xmlNode.SelectSingleNode("businessCentersReference");
            
            if (businessCentersReferenceNode != null)
            {
                if (businessCentersReferenceNode.Attributes["href"] != null || businessCentersReferenceNode.Attributes["id"] != null) 
                {
                    if (businessCentersReferenceNode.Attributes["id"] != null) 
                    {
                        businessCentersReferenceIDRef_ = businessCentersReferenceNode.Attributes["id"].Value;
                        BusinessCentersReference ob = new BusinessCentersReference(businessCentersReferenceNode);
                        IDManager.SetID(businessCentersReferenceIDRef_, ob);
                    }
                    else if (businessCentersReferenceNode.Attributes["href"] != null)
                    {
                        businessCentersReferenceIDRef_ = businessCentersReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        businessCentersReference_ = new BusinessCentersReference(businessCentersReferenceNode);
                    }
                }
                else
                {
                    businessCentersReference_ = new BusinessCentersReference(businessCentersReferenceNode);
                }
            }
            
        
            XmlNode businessCentersNode = xmlNode.SelectSingleNode("businessCenters");
            
            if (businessCentersNode != null)
            {
                if (businessCentersNode.Attributes["href"] != null || businessCentersNode.Attributes["id"] != null) 
                {
                    if (businessCentersNode.Attributes["id"] != null) 
                    {
                        businessCentersIDRef_ = businessCentersNode.Attributes["id"].Value;
                        BusinessCenters ob = new BusinessCenters(businessCentersNode);
                        IDManager.SetID(businessCentersIDRef_, ob);
                    }
                    else if (businessCentersNode.Attributes["href"] != null)
                    {
                        businessCentersIDRef_ = businessCentersNode.Attributes["href"].Value;
                    }
                    else
                    {
                        businessCenters_ = new BusinessCenters(businessCentersNode);
                    }
                }
                else
                {
                    businessCenters_ = new BusinessCenters(businessCentersNode);
                }
            }
            
        
        }
        
    
        #region ExpireRelativeToEvent_
        private CommodityExpireRelativeToEvent expireRelativeToEvent_;
        public CommodityExpireRelativeToEvent ExpireRelativeToEvent_
        {
            get
            {
                if (this.expireRelativeToEvent_ != null)
                {
                    return this.expireRelativeToEvent_; 
                }
                else if (this.expireRelativeToEventIDRef_ != null)
                {
                    expireRelativeToEvent_ = IDManager.getID(expireRelativeToEventIDRef_) as CommodityExpireRelativeToEvent;
                    return this.expireRelativeToEvent_; 
                }
                else
                {
                      return this.expireRelativeToEvent_; 
                }
            }
            set
            {
                if (this.expireRelativeToEvent_ != value)
                {
                    this.expireRelativeToEvent_ = value;
                }
            }
        }
        #endregion
        
        public string expireRelativeToEventIDRef_ { get; set; }
        #region ExpirationDateOffset_
        private DateOffset expirationDateOffset_;
        public DateOffset ExpirationDateOffset_
        {
            get
            {
                if (this.expirationDateOffset_ != null)
                {
                    return this.expirationDateOffset_; 
                }
                else if (this.expirationDateOffsetIDRef_ != null)
                {
                    expirationDateOffset_ = IDManager.getID(expirationDateOffsetIDRef_) as DateOffset;
                    return this.expirationDateOffset_; 
                }
                else
                {
                      return this.expirationDateOffset_; 
                }
            }
            set
            {
                if (this.expirationDateOffset_ != value)
                {
                    this.expirationDateOffset_ = value;
                }
            }
        }
        #endregion
        
        public string expirationDateOffsetIDRef_ { get; set; }
        #region BusinessCentersReference_
        private BusinessCentersReference businessCentersReference_;
        public BusinessCentersReference BusinessCentersReference_
        {
            get
            {
                if (this.businessCentersReference_ != null)
                {
                    return this.businessCentersReference_; 
                }
                else if (this.businessCentersReferenceIDRef_ != null)
                {
                    businessCentersReference_ = IDManager.getID(businessCentersReferenceIDRef_) as BusinessCentersReference;
                    return this.businessCentersReference_; 
                }
                else
                {
                      return this.businessCentersReference_; 
                }
            }
            set
            {
                if (this.businessCentersReference_ != value)
                {
                    this.businessCentersReference_ = value;
                }
            }
        }
        #endregion
        
        public string businessCentersReferenceIDRef_ { get; set; }
        #region BusinessCenters_
        private BusinessCenters businessCenters_;
        public BusinessCenters BusinessCenters_
        {
            get
            {
                if (this.businessCenters_ != null)
                {
                    return this.businessCenters_; 
                }
                else if (this.businessCentersIDRef_ != null)
                {
                    businessCenters_ = IDManager.getID(businessCentersIDRef_) as BusinessCenters;
                    return this.businessCenters_; 
                }
                else
                {
                      return this.businessCenters_; 
                }
            }
            set
            {
                if (this.businessCenters_ != value)
                {
                    this.businessCenters_ = value;
                }
            }
        }
        #endregion
        
        public string businessCentersIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/CommodityRelativeExpirationDates.cpp
// CommodityRelativeExpirationDates.cpp 
#include "CommodityRelativeExpirationDates.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CommodityRelativeExpirationDates::CommodityRelativeExpirationDates(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //expireRelativeToEventNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expireRelativeToEventNode = xmlNode->FirstChildElement("expireRelativeToEvent");

   if(expireRelativeToEventNode){expireRelativeToEventIsNull_ = false;}
   else{expireRelativeToEventIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expireRelativeToEventNode , address : " << expireRelativeToEventNode << std::endl;
   #endif
   if(expireRelativeToEventNode)
   {
      if(expireRelativeToEventNode->Attribute("href") || expireRelativeToEventNode->Attribute("id"))
      {
          if(expireRelativeToEventNode->Attribute("id"))
          {
             expireRelativeToEventIDRef_ = expireRelativeToEventNode->Attribute("id");
             expireRelativeToEvent_ = boost::shared_ptr<CommodityExpireRelativeToEvent>(new CommodityExpireRelativeToEvent(expireRelativeToEventNode));
             expireRelativeToEvent_->setID(expireRelativeToEventIDRef_);
             IDManager::instance().setID(expireRelativeToEventIDRef_,expireRelativeToEvent_);
          }
          else if(expireRelativeToEventNode->Attribute("href")) { expireRelativeToEventIDRef_ = expireRelativeToEventNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expireRelativeToEvent_ = boost::shared_ptr<CommodityExpireRelativeToEvent>(new CommodityExpireRelativeToEvent(expireRelativeToEventNode));}
   }

   //expirationDateOffsetNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationDateOffsetNode = xmlNode->FirstChildElement("expirationDateOffset");

   if(expirationDateOffsetNode){expirationDateOffsetIsNull_ = false;}
   else{expirationDateOffsetIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationDateOffsetNode , address : " << expirationDateOffsetNode << std::endl;
   #endif
   if(expirationDateOffsetNode)
   {
      if(expirationDateOffsetNode->Attribute("href") || expirationDateOffsetNode->Attribute("id"))
      {
          if(expirationDateOffsetNode->Attribute("id"))
          {
             expirationDateOffsetIDRef_ = expirationDateOffsetNode->Attribute("id");
             expirationDateOffset_ = boost::shared_ptr<DateOffset>(new DateOffset(expirationDateOffsetNode));
             expirationDateOffset_->setID(expirationDateOffsetIDRef_);
             IDManager::instance().setID(expirationDateOffsetIDRef_,expirationDateOffset_);
          }
          else if(expirationDateOffsetNode->Attribute("href")) { expirationDateOffsetIDRef_ = expirationDateOffsetNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationDateOffset_ = boost::shared_ptr<DateOffset>(new DateOffset(expirationDateOffsetNode));}
   }

   //businessCentersReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* businessCentersReferenceNode = xmlNode->FirstChildElement("businessCentersReference");

   if(businessCentersReferenceNode){businessCentersReferenceIsNull_ = false;}
   else{businessCentersReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- businessCentersReferenceNode , address : " << businessCentersReferenceNode << std::endl;
   #endif
   if(businessCentersReferenceNode)
   {
      if(businessCentersReferenceNode->Attribute("href") || businessCentersReferenceNode->Attribute("id"))
      {
          if(businessCentersReferenceNode->Attribute("id"))
          {
             businessCentersReferenceIDRef_ = businessCentersReferenceNode->Attribute("id");
             businessCentersReference_ = boost::shared_ptr<BusinessCentersReference>(new BusinessCentersReference(businessCentersReferenceNode));
             businessCentersReference_->setID(businessCentersReferenceIDRef_);
             IDManager::instance().setID(businessCentersReferenceIDRef_,businessCentersReference_);
          }
          else if(businessCentersReferenceNode->Attribute("href")) { businessCentersReferenceIDRef_ = businessCentersReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { businessCentersReference_ = boost::shared_ptr<BusinessCentersReference>(new BusinessCentersReference(businessCentersReferenceNode));}
   }

   //businessCentersNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* businessCentersNode = xmlNode->FirstChildElement("businessCenters");

   if(businessCentersNode){businessCentersIsNull_ = false;}
   else{businessCentersIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- businessCentersNode , address : " << businessCentersNode << std::endl;
   #endif
   if(businessCentersNode)
   {
      if(businessCentersNode->Attribute("href") || businessCentersNode->Attribute("id"))
      {
          if(businessCentersNode->Attribute("id"))
          {
             businessCentersIDRef_ = businessCentersNode->Attribute("id");
             businessCenters_ = boost::shared_ptr<BusinessCenters>(new BusinessCenters(businessCentersNode));
             businessCenters_->setID(businessCentersIDRef_);
             IDManager::instance().setID(businessCentersIDRef_,businessCenters_);
          }
          else if(businessCentersNode->Attribute("href")) { businessCentersIDRef_ = businessCentersNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { businessCenters_ = boost::shared_ptr<BusinessCenters>(new BusinessCenters(businessCentersNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<CommodityExpireRelativeToEvent> CommodityRelativeExpirationDates::getExpireRelativeToEvent()
{
   if(!this->expireRelativeToEventIsNull_){
        if(expireRelativeToEventIDRef_ != ""){
             return boost::shared_static_cast<CommodityExpireRelativeToEvent>(IDManager::instance().getID(expireRelativeToEventIDRef_));
        }else{
             return this->expireRelativeToEvent_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CommodityExpireRelativeToEvent>();
   }
}
boost::shared_ptr<DateOffset> CommodityRelativeExpirationDates::getExpirationDateOffset()
{
   if(!this->expirationDateOffsetIsNull_){
        if(expirationDateOffsetIDRef_ != ""){
             return boost::shared_static_cast<DateOffset>(IDManager::instance().getID(expirationDateOffsetIDRef_));
        }else{
             return this->expirationDateOffset_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<DateOffset>();
   }
}
boost::shared_ptr<BusinessCentersReference> CommodityRelativeExpirationDates::getBusinessCentersReference()
{
   if(!this->businessCentersReferenceIsNull_){
        if(businessCentersReferenceIDRef_ != ""){
             return boost::shared_static_cast<BusinessCentersReference>(IDManager::instance().getID(businessCentersReferenceIDRef_));
        }else{
             return this->businessCentersReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessCentersReference>();
   }
}
boost::shared_ptr<BusinessCenters> CommodityRelativeExpirationDates::getBusinessCenters()
{
   if(!this->businessCentersIsNull_){
        if(businessCentersIDRef_ != ""){
             return boost::shared_static_cast<BusinessCenters>(IDManager::instance().getID(businessCentersIDRef_));
        }else{
             return this->businessCenters_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<BusinessCenters>();
   }
}
}

